packages/expo-updates/ios/EXUpdates/AppLauncher/EXUpdatesAppLauncherWithDatabase.m

```diff
@@ -0,0 +1,199 @@
+#import <EXUpdates/EXUpdatesAppLoaderEmbedded.h>
+#import <EXUpdates/EXUpdatesDatabase.h>
+#import <EXUpdates/EXUpdatesFileDownloader.h>
+#import <EXUpdates/EXUpdatesUtils.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface EXUpdatesAppLauncherWithDatabase ()
+
+@property (nonatomic, strong, readwrite) EXUpdatesUpdate * _Nullable launchedUpdate;
+@property (nonatomic, strong, readwrite) NSURL * _Nullable launchAssetUrl;
+@property (nonatomic, strong, readwrite) NSMutableDictionary * _Nullable assetFilesMap;
+
+@property (nonatomic, strong) EXUpdatesFileDownloader *downloader;
+@property (nonatomic, copy) EXUpdatesAppLauncherCompletionBlock completion;
+
+@property (nonatomic, strong) NSLock *lock;
+@property (nonatomic, assign) int assetsToDownload;
+@property (nonatomic, assign) int assetsToDownloadFinished;
+
+@property (nonatomic, strong) NSError *launchAssetError;
+
+@end
+
+static NSString * const kEXUpdatesAppLauncherErrorDomain = @"AppLauncher";
+
+@implementation EXUpdatesAppLauncherWithDatabase
+
+- (instancetype)init
+{
+  if (self = [super init]) {
+    _lock = [NSLock new];
+    _assetsToDownload = 0;
+    _assetsToDownloadFinished = 0;
+  }
+  return self;
+}
+
++ (EXUpdatesUpdate * _Nullable)launchableUpdateWithSelectionPolicy:(id<EXUpdatesSelectionPolicy>)selectionPolicy
+{
+  EXUpdatesDatabase *database = [EXUpdatesAppController sharedInstance].database;
+  NSError *error;
+  NSArray<EXUpdatesUpdate *>* launchableUpdates = [database launchableUpdatesWithError:&error];
+  if (!launchableUpdates) {
+    NSLog(@"Could not select updates from database: %@", error.localizedDescription);
+    return nil;
+  }
+  return [selectionPolicy launchableUpdateWithUpdates:launchableUpdates];
+}
+
+- (void)launchUpdateWithSelectionPolicy:(id<EXUpdatesSelectionPolicy>)selectionPolicy
+                             completion:(EXUpdatesAppLauncherCompletionBlock)completion
+{
+  NSAssert(!_completion, @"EXUpdatesAppLauncher:launchUpdateWithSelectionPolicy:successBlock should not be called twice on the same instance");
+  _completion = completion;
+  if (!_launchedUpdate) {
+    _launchedUpdate = [[self class] launchableUpdateWithSelectionPolicy:selectionPolicy];
+
+    if (!_launchedUpdate) {
+      _completion([NSError errorWithDomain:kEXUpdatesAppLauncherErrorDomain code:1011 userInfo:@{NSLocalizedDescriptionKey: @"No launchable updates found in database"}], NO);
+      _completion = nil;
+      return;
+    }
+  }
+  
+  _assetFilesMap = [NSMutableDictionary new];
+  NSURL *updatesDirectory = [EXUpdatesAppController sharedInstance].updatesDirectory;
+
+  [_lock lock];
+  if (_launchedUpdate) {
+    for (EXUpdatesAsset *asset in _launchedUpdate.assets) {
+      if ([self ensureAssetExists:asset]) {
+        NSURL *assetLocalUrl = [updatesDirectory URLByAppendingPathComponent:asset.filename];
+        if (asset.isLaunchAsset) {
+          _launchAssetUrl = assetLocalUrl;
+        } else {
+          if (asset.localAssetsKey) {
```

Comment:
It's never assigned to the `assetFilesMap` because it doesn't need to be -- that's for all the other assets that are resolved by the JS (specifically expo-asset) at runtime. The launch asset is accessed separately through the `launchAssetUrl` field.